server/pulp/plugins/rsync/publish.py

```diff
@@ -0,0 +1,443 @@
+        """
+        for t in xrange(10):
+            rv, out = kobo.shortcuts.run(cmd=args, can_fail=True)
+            possible_known_exceptions = \
+                ("ssh_exchange_identification:" in out) or ("max-concurrent-connections=25" in out)
+            if not (rv and possible_known_exceptions):
+                break
+            _logger.info("Connections limit reached, trying once again in thirty seconds.")
+            time.sleep(30)
+        if include_args_in_output:
+            message = "%s\n%s" % (args, out)
+        else:
+            message = out
+        return (rv == 0, message)
+
+    def make_rsync_args(self, files_from, source_prefix, dest_prefix, exclude=None):
+        """
+        Creates a list of arguments for the rsync command
+
+        :param files_from: absolute path to a file with list of file paths to rsync
+        :type files_from: str
+        :param source_prefix: path to directory that contains all paths in files_from
+        :type source_prefix: str
+        :param dest_prefix: path to directory on remote server where files should be rsynced
+        :type dest_prefix: str
+        :param exclude: list of file/directory paths to exclude
+        :type exclude: list
+        :return: list of arguments for rsync command
+        :rtype: list of strings
+        """
+        args = [self._CMD, self._AZQ]
+        if not self.delete:
+            args.extend(["--files-from", files_from, "--relative"])
+
+        if exclude:
+            for x in exclude:
+                args.extend(["--exclude", x])
+        args.extend(self.make_authentication())
+        if self.delete:
+            args.append("--delete")
+        if self.links:
+            args.append("--links")
+        else:
+            args.append("--copy-links")
+        args.append(source_prefix)
+        args.append(self.make_destination(dest_prefix))
+        return args
+
+    def rsync(self):
+        """
+        This method formulates the rsync command based on parameters passed in to the __init__ and
+        then executes it.
+
+        :return: (boolean indicating success or failure, stdout + stderr)
+        :rtype: tuple
+        """
+        if not self.file_list:
+            return (True, "Nothing to sync")
+        output = ""
+        list_of_files = os.path.join(self.get_working_dir(), str(uuid.uuid4()))
+        open(list_of_files, 'w').write("\n".join(sorted(self.file_list)))
+
+        # copy files here, not symlinks
+        (is_successful, this_output) = self.remote_mkdir(self.dest_directory)
+        if not is_successful:
+            _logger.error("Cannot create directory %s: %s" % (self.dest_directory, this_output))
+            return (is_successful, this_output)
+        output += this_output
+        rsync_args = self.make_rsync_args(list_of_files, self.src_directory,
+                                          self.dest_directory, self.exclude)
+        (is_successful, this_output) = self.call(rsync_args)
```

Comment:
There isn't agreement in the Python community about this, but I recommend [exceptions over return codes](http://nedbatchelder.com/text/exceptions-vs-status.html). Specifically this advice is for `call()`. This function checks the return codes many times and then returns, that could all be collapsed into one `except` block for example.

One reason possible not to use Exceptions is if multiple ones could occur and `call()` may need to continue past the first one. I don't think that is the case here, but I could be wrong.